text/0715-construct-upgrade-support.md

```diff
@@ -0,0 +1,1251 @@
+
+At development time, here is how it works for common enterprise scenarios. Assuming developers have upgrade
+their CDK stack code to use new construct. They can use `dry-run` feature which will allow users to
+run the validations on the test/dev account without deploying.
+
+```sh
+$ cdk construct-upgrade --unstable=construct-upgrade --target aws-cdk-lib.aws-dynamodb.TableV2 --dry-run
+Performing construct upgrade validation...
+
+⚠️  Note: This is an experimental feature
+
+Validations:
+✅ No CloudFormation drift detected
+✅ Change set analysis complete
+```
+
+Depending on the deployment workflow and environment constraints, there are two common enterprise
+patterns:
+
+1. CI/CD-driven Deployment:
+For teams with access to CI/CD and production environments, developers commit the upgraded CDK stack
+code to version control. The CI/CD pipeline runs `cdk construct-upgrade` as part of the deployment process.
+Validations are performed automatically before deployment. If any validation fails, the deployment is blocked.
+This approach streamlines deployment, reduces coordination overhead, and ensures migration safety by
+preventing changes from reaching production unless validations pass.
+
+2. Ops-led Deployment:
+The development team upgrades the source code and runs validations in a dev/test environment using `--dry-run`.
+The development team provides the source code with validation output to the operations team. The ops team
+reviews the validation output, runs the migration in each production region, and deploys the changes.
+Once migration is complete, the development team commits and pushes the updated source code to trigger CI/CD,
+which is an no-op at this point.
+
+Alternatively, the development team can upgrade the source code, commit and push the changes with CI/CD pipelines
+deployment temporarily disabled. The operations team runs the migration in each production region and validates
+deployment success. The development team re-enables CI/CD to roll out the changes through the normal pipeline.
+
+**Known Limitation**: CI/CD-driven deployment has limitations when using pipeline systems that don't easily allow
+command configurations (e.g., CDK L3 Pipeline construct where deployments rely on CloudFormation actions rather
+than direct CDK CLI commands).
+
+In these systems, implementing the CI/CD-driven deployment would require:
+
+- Blocking the CI system while adding the `cdk construct-upgrade` command
+- Committing the change to trigger the upgrade validation
+- Unblocking the CI system and preventing new commits during the process
+- Removing the upgrade command from CI pipeline configuration after completion
+
+This process is cumbersome and typically requires administrative permissions to modify the CI system. Given
+these challenges, we recommend using Ops-led deployment for teams with complex pipeline setups until a more
+elegant solution is developed in the future updates.
```

Comment:
I think it's really important to define _complex_ here. It would also be good to understand how many customers have complex setups i.e. does the proposed approach actually work for a majority of customers?

Reply:
We don't currently have a way to track the analytics usage of users' CI/CD pipeline setup unfortunately. That's why we provided a workaround approach for a less-ideal solution.